Ci deletedtxt

kba opened the pull request:
A simple approach that should work for PR

kba committed:
ci: check translators for reused IDs

kba committed:
ci: check translators for reused IDs

kba committed:
ci: check for deleted.txt consistency

kba committed:
ci: check for deleted.txt consistency

zuphilip commented on .ci/package.json:
```diff
@@ -0,0 +1,19 @@
+  "main": "index.js",
+  "scripts": {
+    "test": "export SKIP_WARN=true ; find .. -maxdepth 1 -name '*.js' -print0|xargs -0 -n1 -P8 ./checkTranslator.sh 2>/dev/null; ./checkDeletedTxt.sh",
+    "start": "find .. -maxdepth 1 -name '*.js' -print0|xargs -0 -n1 -P1 ./checkTranslator.sh; ./checkDeletedTxt"
```
It looks like the `.sh` is missing here at the end. But we could also just call the `checkDeletedTxt.sh` script as a second script in travis. No?

zuphilip commented on .ci/checkTranslator.sh:
```diff
@@ -40,19 +45,28 @@ notOneTranslatorID () {
-      duplicateIds=$(grep -r '"translatorID"' "$dir"/*.js \
-              | sed -e 's/[" ,]//g' -e 's/^.*://g' \
-              | sort | uniq -d)
+      duplicateIds=$(grepTranslatorId "$dir"/*.js | sort | uniq -d)
```
Actually, this will be performed for each translator, which seems to be then too much, because we already know the specific `id` to look for. I would suggest something like
```bash
duplicateIds=$(grep "$id" | uniq -d)
```
which can then be tested to being empty.

zuphilip commented on .ci/checkTranslator.sh:
```diff
@@ -0,0 +1,163 @@
+    local dir id duplicateIds
+    dir=$(dirname "$TRANSLATOR")
+    id=$(grep -r '"translatorID"' "$TRANSLATOR" \
+        | sed -e 's/[" ,]//g' -e 's/^.*://g')
```
Isn't this the newly introduced function:
```
$(grepTranslatorId "$TRANSLATOR")
```
?

zuphilip replied:
Yes, seems so.

zuphilip commented on .ci/checkTranslator.sh:
```diff
@@ -0,0 +1,163 @@
+    dir=$(dirname "$TRANSLATOR")
+    id=$(grepTranslatorId "$TRANSLATOR")
+    if [[ -n "$id" ]];then
+      duplicateIds=$(grepTranslatorId "$dir"/*.js | sort | uniq -d)
```
Actually, this will be performed for each translator, which seems to be then too much, because we already know the specific `id` to look for. I would suggest something like
```bash
duplicateIds=$(grep "$id" | uniq -d)
```
which can then be tested to being empty.

zuphilip commented on .ci/checkDeletedTxt.sh:
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
+
+SCRIPT_DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
+. "$SCRIPT_DIR/helper.sh"
```
Is this `.` doing the same as `source "$SCRIPT_DIR/helper.sh"` in the other file? I would rather prefer the latter one for better understanding what is done and being more consistent.

zuphilip commented on .ci/checkDeletedTxt.sh:
```diff
@@ -0,0 +1,38 @@
+        for f in "${deletions[@]}";do
+            local id=$(git show $MASTER:"$f"|grepTranslatorId)
+            if ! grep -F "$id" '../deleted.txt';then
+                echo "not ok - $id ($f) should be added to deleted.txt"
```
We can also use colors here, i.e.
```bash
echo "${color_notok}not ok${color_reset} - ..."
```

zuphilip commented on .ci/checkDeletedTxt.sh:
```diff
@@ -0,0 +1,38 @@
+        curVersion=$(head -n1 "../deleted.txt"|grep -o '[0-9]\+')
+        origVersion=$(git show "$MASTER:deleted.txt"|head -n1|grep -o '[0-9]\+')
+        if (( curVersion <= origVersion ));then
+            echo "not ok - version in deleted.txt needs to be increased"
```
We can also use colors here, i.e.
```bash
echo "${color_notok}not ok${color_reset} - ..."
```

zuphilip commented on .ci/checkDeletedTxt.sh:
```diff
@@ -0,0 +1,38 @@
+    if (( ${#deletions[@]} > 0 ));then
+        for f in "${deletions[@]}";do
+            local id=$(git show $MASTER:"$f"|grepTranslatorId)
+            if ! grep -F "$id" '../deleted.txt';then
```
We only should look here at delete translators and ignore other files which might get deleted. Thus, I suggest to add here a test statement for the following three lines:
```bash
if [[ -n "$id" ]];then
```

zuphilip replied:
Let us make the `grep` silent with `-q`.

zuphilip commented on .ci/checkDeletedTxt.sh:
```diff
@@ -0,0 +1,38 @@
+            (( failed += 1))
+        fi
+    fi
+    exit $failed
```
Let us also add an `okay` statement if everything works okay.

zuphilip committed:
Improve checkDeletedTxt.sh a little

zuphilip commented:
Okay, I did now some little changes according to my comments above. @kba What do you think?

zuphilip committed:
Add multistr option to jshintrc

zuphilip commented:
Okay, I just rebase the branch to be compatible with latest changes of the master. Everything looks good for me, but on Travis I see:
```
$ ./checkDeletedTxt.sh

fatal: ambiguous argument 'master': unknown revision or path not in the working tree.

Use '--' to separate paths from revisions, like this:

'git <command> [<revision>...] -- [<file>...]'

ok - deleted.txt

The command "./checkDeletedTxt.sh" exited with 0.
```

Any ideas @kba? One of the first lines on travis is the checkout the current branch (but maybe we need more?):
```
git clone --depth=50 --branch=ci-deletedtxt https://github.com/infolis/translators.git infolis/translators
```

zuphilip commented:
Eh... only the travis on infolis shows this message: https://travis-ci.org/infolis/translators/builds/194249781 but not the travis here https://travis-ci.org/zuphilip/translators/builds/194249804 . What? 😕

kba committed:
Merge branch 'ci-deletedtxt', only check translator JS for deletion

kba committed:
rev-parse the HEAD revision without abbrev-ref

zuphilip committed:
.ci/checkDeletedTxt.sh: Warn/exit if master not available

Actually, this script will either be used in a local feature branch
and then we can assume that master branch is also available, or
in the Travis check of the official repo, where in each PR the
master branch is available.

In all other cases we skip now the test and show just a warning
that we are doing that.

zuphilip commented:
Okay, there is a difference in the two travis checks (push vs. pull request) because checking out the repo is done differently:
```
$ git clone --depth=50 --branch=ci-deletedtxt https://github.com/infolis/translators.git infolis/translators
$ cd infolis/translators
$ git checkout -qf 62ba1ef790ccfc735de6e6c10927e597594c5c1f
```
vs
```
$ git clone --depth=50 https://github.com/zuphilip/translators.git zuphilip/translators
$ cd zuphilip/translators
$ git fetch origin +refs/pull/2/merge:
$ git checkout -qf FETCH_HEAD
```
However, we can concentrate on pull request, because that is important to work. Also in local settings it should work, because then `master` branch should always be there. For other cases, I switched now to a warning and skip the test.

zuphilip committed:
CI: Update Readme, use stable path for jshintrc

zuphilip committed:
Add option to call checkTranslator.sh with label only

zuphilip commented:
This is now merged upstream in the official repo, see https://github.com/zotero/translators/pull/1238. Thank you very much!
